File: Tutorial-1/Baseline/baseline.cpp
#include <iostream>
#include <chrono>
#include <vector>

std::vector<size_t> create_array(size_t size) {
    std::vector<size_t> arr;
    arr.reserve(size);
    for (size_t i = 0; i < size; ++i) {
        arr.push_back(i * i);
    }
    return arr;
}

int main() {
    const size_t NUM_ARRAYS = 1000000;
    const size_t ARRAY_SIZE = 10;

    auto start = std::chrono::high_resolution_clock::now();

    for (size_t i = 0; i < NUM_ARRAYS; ++i) {
        std::vector<size_t> my_array = create_array(ARRAY_SIZE);
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> duration = end - start;

    std::cout << "C++ Time taken: " << duration.count() << " seconds" << std::endl;

    return 0;
}

File: Tutorial-1a/Baseline/baseline.rs
use std::time::Instant;

fn create_array(size: usize) -> Vec<usize> {
    let mut arr = Vec::with_capacity(size);
    for i in 0..size {
        arr.push(i * i);
    }
    arr
}

fn main() {
    const NUM_ARRAYS: usize = 1000000;
    const ARRAY_SIZE: usize = 10;

    let start = Instant::now();

    for _ in 0..NUM_ARRAYS {
        let _my_array = create_array(ARRAY_SIZE);
    }

    let duration = start.elapsed();

    println!("Rust Time taken: {:.6} seconds", duration.as_secs_f64());
}
